components/service/glean/src/main/java/mozilla/components/service/glean/Dispatchers.kt

```diff
@@ -11,31 +11,61 @@ import kotlinx.coroutines.ObsoleteCoroutinesApi
 import kotlinx.coroutines.launch
 import kotlinx.coroutines.newSingleThreadContext
 import kotlinx.coroutines.runBlocking
+import mozilla.components.support.base.log.logger.Logger
+import java.util.concurrent.ConcurrentLinkedQueue
 
 @ObsoleteCoroutinesApi
 internal object Dispatchers {
     class WaitableCoroutineScope(val coroutineScope: CoroutineScope) {
         // When true, jobs will be run synchronously
         internal var testingMode = false
 
+        // When true, jobs will be queued and not ran until
+        internal var queueTasks = true
+
+        // Use a [ConcurrentLinkedQueue] to take advantage of it's thread safety and no locking
+        internal val taskQueue: ConcurrentLinkedQueue<() -> Unit> = ConcurrentLinkedQueue()
+
+        private val logger = Logger("glean/Dispatchers")
+
+        companion object {
+            const val MAX_QUEUE_SIZE = 100
+        }
+
         /**
-         * Launch a block of work asyncronously.
+         * Launch a block of work asynchronously.
          *
          * If [setTestingMode] has enabled testing mode, the work will run
          * synchronously.
          *
          * @return [Job], or null if run synchronously.
          */
+        @Suppress("LongMethod")
         fun launch(
             block: suspend CoroutineScope.() -> Unit
         ): Job? {
-            return if (testingMode) {
-                runBlocking {
-                    block()
+            return when {
+                testingMode -> {
```

Comment:
What if we wrap this somehow and make sure that enqueued tasks are also `testingMode` aware? This will make it more clearer in tests too (see my comments below)